Allow disabling of content listener.

benglass opened the pull request:
We are working on integrating this into our CMS but would like to use our own content listener.

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes

## TO DO 

- [X] Add enable_content_listener boolean node
- [X] Update XML schema
- [X] Add tests
- [ ] Update docs

dbu commented:
looks sane to me, and makes sense.

the only thing missing imo is a PR to adjust the doc, with a `.. since` to the configuration reference of the seo bundle.

lsmith77 commented:
cc @ElectricMaxxx

WouterJ commented on DependencyInjection/Configuration.php:
```diff
@@ -85,6 +85,9 @@ public function getConfigTreeBuilder()
                     ->end()
                 ->end()
+                ->arrayNode('content_listener')
+                    ->canBeDisabled()
```
Instead, I think we should use a `booleanNode` and call it `use_content_listener`

dbu replied:
also possible. but if we ever want to have other configuration on content_listener, it will be unelegant. then again that is not too likely, so +1 for use_content_listener (without a content_listener array node)

benglass replied:
This is actually how I originally wrote it (booleanNode) however my thought was we may want to allow specifying a service_id to use instead of the out of the box content listener so you could provide your own content listener and not have to rewrite the alternate locale provider portion. I'm perfectly happy to switch it to a booleanNode, just figured I'd provide my rational first

WouterJ replied:
While we agree, I just wanted to point out that you can easily keep BC by keeping content_listener.enabled and use_content_listener in sync in case we want to add more config in the future.

WouterJ replied:
Hmm, what about a content_listener setting (scalarNode) which could either be the service id or false/null to disable it?

benglass replied:
The default value would be true which would cause it to use the existing one?

WouterJ replied:
Default value would be thre service id of thre drfault listener.

benglass replied:
I like this its more explicit

dbu replied:
does the bundle need to know the service id of the default listener? if yes that would be the right thing to do. although, would that mean that if its the default value, SeoBundle loads the service and otherwise its assumes the application provides the service somewhere else? and, does the SeoBundle need to know the name of that listener, does it inject it elsewhere?

benglass replied:
@dbu I think the way this is handled by FOSUserBundle where you can override the service id for the user manager is that the default service's id is like cmf_seo.content_listener_default and the dependency injection extension sets up cmf_seo.content_listener as an alias based on the configuration

dbu replied:
alright, sounds like a plan!

WouterJ commented:
The XML schema needs to be updated too

ElectricMaxxx commented:
I think there a still some features missing in the xml schema :-(

ElectricMaxxx commented:
@benglass looks good to me,but can you add a test for both situations. Should be something like 
https://github.com/symfony-cmf/SeoBundle/blob/master/Tests/Unit/DependencyInjection/CmfSeoExtensionTest.php#L119-L122

dbu commented:
@benglass any chance to wrap this up? we think about tagging a new minor version of SeoBundle this week to get the sitemap generation feature in a stable release.

ElectricMaxxx commented on Tests/Unit/DependencyInjection/CmfSeoExtensionTest.php:
```diff
@@ -121,11 +121,13 @@ public function testAlternateLocaleWithPhpcr()
-            'setAlternateLocaleProvider',
-            array($this->container->getDefinition('cmf_seo.alternate_locale.provider_phpcr'))
-        );
+        if ($this->container->has('cmf_seo.event_listener.seo_content')) {
```
Wyh are you doing this?

benglass replied:
Because this PR adds the ability to remove the seo content listener, therefore this test needs to only assert that the event listener exists with this method call if the event listener exists

ElectricMaxxx replied:
normally the alternate locale provider is injected into the SeoPresentation, when activated and with phpcr as implementation, this is done inside of the `CmfSeoExtension`. I would propose to add your changes there.

ElectricMaxxx replied:
Do then you should use the config section above of the test and check for the existence by on or  many asserts. Cause i think your/my event listener will exist only, when there is a line in the configuration, right?

benglass commented on DependencyInjection/CmfSeoExtension.php:
```diff
@@ -81,8 +81,17 @@ public function load(array $configs, ContainerBuilder $container)
 
-        if ($this->isConfigEnabled($container, $config['alternate_locale'])) {
-            $this->loadAlternateLocaleProvider($config['alternate_locale'], $container);
+        if ($config['content_listener']) {
```
@dbu @WouterJ @ElectricMaxxx 

This is the current state of the logic.

The default value of content_listener is cmf_seo.event_listener.seo_content_default
If this is the value of the config key then we load the content listener service definition (which is a tagged service, we dont want to do load file unless this is the listener in use or we would have multiple kernel request listeners)
If the user provides another service id, they are responsible for tagging it at as a kernel request listener. We could do this automatically but I believe it would have to be in compiler pass because the service may be from another bundle. I think we may be best off just allowing disabling of the content listener and provide doc instructions that you can just provide your own. The other thing about this PR is that the loadAlternateLocaleProvider functionality is only applied to the default seo content listener so the user would have to handle this in their service if they disable this listener.

At this point I am leaning towards just changing 'content_listener' to 'enable_content_listener' as a boolean and providing instructions for users to provide their own listeners. It reduces code complexity here.

WouterJ replied:
Yeah, I thought we did things with the service that was configured as content listener. I agree with the tagging.

+1 for a booleanNode now

benglass commented:
@dbu @WouterJ @ElectricMaxxx 

I have now added tests and updated the xml schema. I do not have experience working with xsd files so can someone verify I did it correctly? The current schema file does not seem up to date with the other config values but I think you may have already fixed this in the master branch?

Doc PR will be coming but it would be great to get this merged in for a numbered release

WouterJ commented on Resources/config/schema/seo-1.0.xsd:
```diff
@@ -39,4 +39,8 @@
         </xsd:restriction>
     </xsd:simpleType>
+
+    <xsd:simpleType name="enable_content_listener">
```
these changes should be removed and instead, you should add this between line 20 and 21 in this file:

````xsd
<xsd:attribute name="enable-content-listener" type="xsd:boolean" default="true" />
````

WouterJ commented:
Thanks @benglass! Apart from my XSD comment, it looks great.

Could you please also rebase this PR? Then Travis can run a build and we can merge if Travis is happy :)

benglass commented:
@WouterJ Squashed, rebase and awaiting tests

benglass commented:
@WouterJ @ElectricMaxxx @dantleech Tests passing ready for merge

WouterJ commented:
:+1:

dbu commented:
thanks a lot! created an issue so we don't forget the documentation: https://github.com/symfony-cmf/symfony-cmf-docs/issues/644
